lich.rbw

```diff
@@ -5186,14 +5186,14 @@ main_thread = Thread.new {
         launcher_cmd = launcher_cmd.tr('/', "\\") if (RUBY_PLATFORM =~ /mingw|win/i) and (RUBY_PLATFORM !~ /darwin/i)
       end
       begin
-        if custom_launch_dir
+        unless custom_launch_dir.empty?
```

Comment:
Can someone check this and make sure it's correct? It works for me, but I'm not sure if custom_launch_dir can be null and if there would be an issue in that case.

Reply:
`custom_launch_dir` can be nil. You could just have the check be `unless (custom_launch_dir.empty? ||  custom_launch_dir.nil?` ?

Reply:
ok, I added the nil? check. Pretty sure the `nil?` needs to come before the `empty?`

Reply:
If they're in parentheses it doesn't matter.

Reply:
I don't think that is correct. I've read/written very little Ruby, but I don't know another language that works like that and my simple test says ruby doesn't either. This doesn't work:
```ruby
foo = nil
if (foo.empty? || foo.nil?)
  puts "works"
end
```
The parentheses do nothing. If you reverse the order, this works:
```ruby
foo = nil
if foo.nil? || foo.empty?
  puts "works"
end
```
It works as expected. The `||` short circuits and only evaluates the `nil?`. If you put them in the other order, the `empty?` is executed first and gives an undefined method error.

I don't think the parentheses do anything in this case, because there's no ambiguity to the order of operations.

Reply:
Note: in Ruby, `and` and `&&` are not the same, and neither are `or` or `||'. The difference is operator precedence. See: https://mixandgo.com/learn/ruby/boolean-operator-precedence

I'm not familiar with the code, here, but if we're structuring our `||` check with short-circuiting governing the success of the check, that suggests we're unsure of the data type for `custom_launch_dir`. Is that the case? Because a `.empty` check on `nil` doesn't make sense in Ruby. Is `custom_launch_dir` initialized as a string?

Reply:
custom_launch_dir is the either a string or nil. It's initialized at line 4604 `if custom_launch_dir = @launch_data.find { |opt| opt =~ /CUSTOMLAUNCHDIR=/ }`

Reply:
If the variable can be `nil` or a string, I think your check works. If I were writing the original code, I'd initialize it such that I knew the data type. But I'm not an expert.

Reply:
Damn, thanks for the lesson. That makes sense that the `.empty?` is undefined if the variable is nil.

Reply:
Well, now's as good a time as any. How would you write the initialization? We're already messing with this code, so we might as well make it easy to read while we're at it. Let's ignore for a moment that I had to refer to this line by a number greater than 4000.

On another topic, is anyone planning to refactor this to make it more readable/maintainable? Having 2 huge source files containing almost all of the logic is awkward. I'm really an outsider here. It's not obvious where any discussion/planning happens.

Reply:
Now is a good time to refactor this. Making it its own file like you have here is good. We can get it tested easy enough, then it can go into Staging for general testing. It shouldn't impact a lot of people... not too many linux users around.